docs/concepts/validators.md

```diff
@@ -1,635 +1,575 @@
+      x: int
+
+      @field_validator('x', mode='after')
+      @classmethod
+      def validate_x(cls, v: int) -> int:
+          if v % 42 == 0:
+              raise PydanticCustomError(
+                  'the_answer_error',
+                  '{number} is the answer!',
+                  {'number': v},
+              )
+          return v
+
+
+  try:
+      Model(x=42 * 2)
+  except ValidationError as e:
+      print(e)
+      """
+      1 validation error for Model
+      x
+        84 is the answer! [type=the_answer_error, input_value=84, input_type=int]
+      """
+  ```
+
+## Validation info
+
+Both the field and model validators callables (in all modes) can optionally take an extra
+[`ValidationInfo`][pydantic.ValidationInfo] argument, providing useful extra information, such as:
+
+- [already validated data](#validation-data)
+- [user defined context](#validation-context)
+- the current validation mode: either `'python'` or `'json'` (see the [`mode`][pydantic.ValidationInfo.mode] property)
+- the current field name (see the [`field_name`][pydantic.ValidationInfo.field_name] property).
+
+### Validation data
+
+For field validators, the already validated data can be accessed using the [`data`][pydantic.ValidationInfo.data]
+property. Here is an example than can be used as an alternative to the [*after* model validator](#model-after-validator)
+example:
+
+```python
+from pydantic import BaseModel, ValidationInfo, field_validator
+
+
+class UserModel(BaseModel):
+    password: str
+    password_repeat: str
+    username: str
+
+    @field_validator('password_repeat', mode='after')
+    @classmethod
+    def check_passwords_match(cls, value: str, info: ValidationInfo) -> str:
+        if value != info.data['password']:
+            raise ValueError('Passwords do not match')
+        return value
+```
 
-This may be undesirable if, for example, you want to define a validator to validate fields that will only be present
-on subclasses of the model where the validator is defined.
+!!! warning
+    As validation is performed in the [order fields are defined](./models.md#field-ordering), you have to
+    make sure you are not accessing a field that hasn't been validated yet. In the code above, for example,
+    the `username` validated value is not available yet, as it is defined *after* `password_repeat`.
 
-If you want to disable these checks during class creation, you can pass `check_fields=False` as a keyword argument to
-the validator.
+The [`data`][pydantic.ValidationInfo.data] property is `None` for [model validators](#model-validators).
 
-## Validation Context
+### Validation context
 
-You can pass a context object to the validation methods which can be accessed from the `info`
-argument to decorated validator functions:
+You can pass a context object to the [validation methods](./models.md#validating-data), which can be accessed
+inside the validator functions using the [`context`][pydantic.ValidationInfo.context] property:
```

Comment:
Maybe add a note about runtime / dynamic data here

Reply:
What do you mean by *runtime / dynamic data* exactly?

Reply:
Ignore me, I don't think this is actually helpful to clarify here.